### DISCUSSION TITLE
How to open IV reference stream for smart amp use case in linux (SOF-Driver)

### DISCUSSION THREAD

**Discussion by Venkata-Prasad-Potturu** [thesofproject/sof#8082] [Q&A] [answered]
We are trying to open an IV reference stream when we start playback on Smart amp speaker. 
Found a reference to open a IV reference stream from alsa ucm files in CRAS (chromium repo).
Taken sample ucm example from cras ucm as below and used the same in Linux machines (Pulse audio and Pipewire) , but IV reference stream is not opening.
Could you please help me to understand how to open IV reference stream of Smart amp for playback use case.
Thanks :)

**HiFi.conf**
SectionDevice."Speaker".0 {
	Comment "Speaker"
	Value {
		PlaybackPriority 100
		PlaybackPCM "hw:sofnau8821max,1"
		EchoReferenceDev "Echo Reference"
	}
	EnableSequence [
		cdev "hw:sofnau8821max"
		exec "echo Speaker EnableSequence"
	]
	DisableSequence [
		cdev "hw:sofnau8821max"
		exec "echo Speaker DisableSequence"
	]
}
SectionDevice."Echo Reference".0 {
	Value {
		CapturePCM "hw:sofnau8821max,1"
	}
	EnableSequence [
		exec "echo Speaker reference capture path EnableSequence"
	]
	DisableSequence [
	]
}

**Comment by Venkata-Prasad-Potturu** [comment]
@plbossart @ranj063 
Please comment on this.

**Reply by lgirdwood** [reply]
@Venkata-Prasad-Potturu are you trying to tell CRAS which PCM is to be opened for IV ?

**Comment by Venkata-Prasad-Potturu** [comment]
@lgirdwood 
Not CRAS, In our case we are using Pulseaudio and to open IV PCM stream we have mentioned in UCM file as above.
Is this correct way to open IV stream when playback starts on speaker.

**Reply by lgirdwood** [reply]
I think Pulseaudio is missing and unlikely to acquire SmartAmp support. I would reach out to pipewire developers for support in Pipewire.

**Reply by cujomalainey** [reply]
Indeed @Venkata-Prasad-Potturu EchoReferenceDev is a CRAS concept. I sent the patches definitions upstream, if they land pulse/pipewire should be more accepting of the definitions once implemented

**Reply by plbossart** [reply]
ironically it should be easier to upstream if this is not perceived as an Intel-only extension :-)

**Comment by plbossart** [comment]
Are we confusing echo reference and IV feedback by any chance? It's possible that for test purposes the IV feedback is provided to userspace, but the intended use for the IV feedback is to be used by a speaker protection module inside the DSP firmware.

**Reply by bhiregoudar** [reply]
We are looking for IV feedback to be provided to speaker protection module inside DSP firmware. We need info on how capture is invoked when playback is started.

**Reply by plbossart** [reply]
it's a miss in DPCM at the moment, you need to have an application start the capture stream. It's the same for echo reference. That's why you see these UCM parts, so that when playback is enabled you also start the echo reference stream.

**Reply by vijendarmukunda** [reply]
@plbossart :
 As per our understanding mentioning EchoReferenceDev  in ALSA UCM file is not enough to invoke the capture device open when speaker pcm device gets opened in pipewire/pulse audio environment as it has to handle EchoReferenceDev  reference similar to CRAS. 
 Please correct me, if i am wrong. 
 As you mentioned it's miss in DPCM, but still CRAS able to open IV reference device when speaker pcm device gets opened.
 If We understood Correctly, we need to have similar sequence to open IV stream pcm device from Pipewire/PulseAudio environment.

**Reply by plbossart** [reply]
questions on CRAS and UCM should be directed to @cujomalainey and @bzhg

**Reply by cujomalainey** [reply]
Not sure I understand entirely the question but I will give context to hopefully answer.

EchoReferenceDev is indeed a CRAS only concept. I tried to upstream the definition a few years ago but upstream didn't seem super interested.

Usually we don't actually use the echo in CRAS, the reason we see typically in the default case the reference arrives after the echo which just makes the AEC mad.

Regarding IV, the only reason CRAS opens it is because SOF can't do it itself. We have no need for the data.

**Reply by felakuti4life** [reply]
I'm working with @bhiregoudar and @vijendarmukunda a bit on this.

Just to summarize, the gist of what's happening here is that the IV signal is a PCM device that needs to have a stream opened in lockstep with the speaker device for speaker protection purposes. I believe the analogy in AEC is that the echo reference PCM device used to derive the error signal for AEC in CRAS is enabled and disabled in lockstep with the speaker in a similar way.

 Given this (and apologies if this is a naive interpretation), it seems like the best path is coupling the IV device enable / disable into the speaker verb EnableSequence/DisableSequence in the sound card's UCM config? While this could be done with pipewire-media-session / wireplumber, having a UCM verb that can still be targeted by any linux audio management upstream of alsa seems a bit more portable for more distros. Is there anything here that is preventing the IV device enable / disable from being set via the UCM verb?

**Reply by cujomalainey** [reply]
@felakuti4life that is essentially what we have done with [EchoReferenceDev](https://github.com/chromeos/adhd/blob/main/cras/src/server/cras_alsa_ucm.c#L45)

It looks like i am [remembering wrong](https://mailman.alsa-project.org/hyperkitty/list/alsa-devel@alsa-project.org/message/JAMDBIYOPKKC7WZUOGZY7ZERTX5HURRM/) and upstream didnt have any issues with it. I will send the patch again and see what they say.

**Reply by cujomalainey** [reply]
One thing missing in the definition is the difference between a proper reference versus an IV feedback stream that needs to be opened but we don't actually care about.

**Reply by cujomalainey** [reply]
Patches sent again https://mailman.alsa-project.org/hyperkitty/list/alsa-devel@alsa-project.org/thread/JFOBKRMOJZRQAG4D6MNC2SYIMY7BHQO2/

**Reply by felakuti4life** [reply]
@cujomalainey Thanks for the clarification, this looks like exactly what would be needed. `EchoReferenceDev` is just in this case a generic coupled device.

**Reply by bhiregoudar** [reply]
It seems along with the changes provided by Curtis, pipewire/pulseaudio changes are also required to support the IV feedback path.
@felakuti4life @lgirdwood : Could you please take this forward for pipewire/pulseaudio?

**Reply by cujomalainey** [reply, marked as answer]
To close out this thread, it appears like it was a common misunderstanding of how echo refs are properly done in also. See Jaroslav's responses.
